### PR TITLE
Fixes #10535 - building a token instead of creating it until host is saved

### PR DESCRIPTION
**Pull request by orrabin**

### CODE REVIEW

**Comment by orrabin** [comment]
@ohadlevy I tried to create a test for this but couldn't reproduce it without using discovery, I'll open up a PR for discovery with a test soon

**Comment by domcleal** [comment]
Note that this is against an old, closed Discovery bug but should be a new, open, Foreman bug.

**Comment by orrabin** [comment]
@domcleal the discovery bug wasn't properly fixed and I wasn't sure if it makes sense to open up a foreman bug if I can't reproduce it without discovery

**Comment by domcleal** [comment]
We'll still need it to track the release this is shipped in, and to even describe what/why this is.

File: app/models/host/managed.rb [hunk 1]
```diff
@@ -213,7 +213,7 @@ def clear_data_on_build
 
   def set_token
     return unless Setting[:token_duration] != 0
-    self.create_token(:value => Foreman.uuid,
+    self.build_token(:value => Foreman.uuid,
                       :expires => Time.now.utc + Setting[:token_duration].minutes)
```
Comment (domcleal) [inline app/models/host/managed.rb]: line up here

**Comment by lzap** [comment]
Okay I have tested this with Discovery and without discovery. Looks good. Would folks mind that this is merged without test (which is actually part of discovery - referenced above)?

**Comment by domcleal** [comment]
Why isn't it possible to write a unit test here?

**Comment by lzap** [comment]
Yeah small unit test is possible, functional test would be more challenging but can be stubbed too.

**Comment by orrabin** [comment]
@domcleal every test I wrote passed even before the fix.
I was trying to do something like:

``` ruby
 test "token is not created until host is saved" do
      Setting[:token_duration] = 30
      h=FactoryGirl.build(:host, :managed, :operatingsystem => nil)
      h.save
      assert Token.where(:host_id => h.id).empty?
  end
```

so we get to set_token after validations but the host isn't saved yet.

**Comment by lzap** [comment]
Nice, can we do the positive case? I am unable to find such a test. Just save and check if a token was associated.

**Comment by lzap** [comment]
Or if we have that already, no need to write other one.

**Comment by domcleal** [comment]
I understand the patch, but still don't quite understand the ticket at http://projects.theforeman.org/issues/10535.  You're starting with Host::Discovered, it .becomes(Host::Managed), fails validation, reverts to Host::Discovered and a token is committed to the database, or..?

**Comment by orrabin** [comment]
@domcleal you are exactly right. The "reverts to Host::Discovered" is a result of canceling the provision. The steps would be:
1. discover a host
2. click provision
3. don't fill in anything and click save
4. after validations are shown click cancel
5. go to db and do Token.where(:host_id => ...)
There should not be a token but there is.

**Comment by domcleal** [comment]
I think you should be able to recreate that in a test then, see test/unit/orchestration_test.rb for instance which uses another subclass, Host::Test.

File: test/unit/orchestration_test.rb [hunk 2]
```diff
@@ -29,6 +29,27 @@ def setTest; true; end
     end
   end
 
+  test "test tokens are not created until host is saved" do
+    class Host::Test < Host::Base
+
```
Comment (lzap) [inline test/unit/orchestration_test.rb]: Rubocop will not like this empty line I think...

File: test/unit/orchestration_test.rb [hunk 3]
```diff
@@ -29,6 +29,27 @@ def setTest; true; end
     end
   end
 
+  test "test tokens are not created until host is saved" do
```
Comment (ohadlevy) [inline test/unit/orchestration_test.rb]: why is this in orchestration_test?

Comment (orrabin) [inline test/unit/orchestration_test.rb]: by mistake, I moved it before seeing your comment :)

**Comment by orrabin** [comment]
@domcleal ty, I added a test that reproduced it in develop

**Comment by lzap** [comment]
If here are no objections, I am going to merge once tests are confirmed green...

**Comment by lzap** [comment]
Merged as c2ff4bd6874155459a51af070abb381fd59be262, thank you!

**Comment by domcleal** [comment]
One further thought after this has been rushed in, is that https://github.com/theforeman/foreman_discovery/commit/52cd340ad7cf31fc54948bfca844b162430c7f40 (and by extension, https://github.com/theforeman/foreman_discovery/pull/180), ought to be in Foreman, not in Discovery since this is a Foreman bug.  Thoughts?

**Comment by orrabin** [comment]
@domcleal I'm not sure it's that easy to reproduce in foreman as in discovery but we should create a migration for all hosts that have more then one token to delete it.

**Comment by domcleal** [comment]
Yes, I agree that it depends on Discovery to have happened, but currently the two fixes may not work in parallel - if the Discovery release is made with that migration, and this Foreman fix isn't released, then it's going to prematurely tidy the database only for more tokens to be generated.

I don't think that Discovery should be trying to add migrations around Host::Managed itself.

**Comment by lzap** [comment]
You are right, absolutely. The context here is we struggled really hard to find real cause, we pushed one fix, it did not work in all use cases which we found after this was shipped downstream and after that failed QA we realized that this is actually Foreman Core bug. 

I made a comment in the issue to copy the migration over to Core if we decide to backport. Let's wait if more users hit this. We only seen one report so far. Opinions?

**Comment by domcleal** [comment]
I don't understand why the migration would only move if used in 1.8, surely the opposite is more true.  Anyway, I filed #10558 (also noting that an index should be added).

**Comment by ohadlevy** [comment]
:+1:

**Pull request closed without merging**